[LoopVectorize][NFC] Rewrite tests to check output of vplan cost model

david-arm opened the pull request:
Currently it's very difficult to improve the cost model for tail-folded
loops because as soon as you add a VPInstruction::computeCost function
that adds the costs of instructions such as VPInstruction::ActiveLaneMask
and VPInstruction::ExplicitVectorLength the assert in
LoopVectorizationPlanner::computeBestVF fails for some tests. This is
because the VF chosen by the legacy cost model doesn't match the vplan
cost model. See PR #90191. This assert is currently making it difficult
to improve the cost model.

Hopefully we will be in a position to remove the assert soon, however
in order to do that we have to fix up a whole bunch of tests that rely
upon the legacy cost model output. I've tried my best to update
these tests to use vplan output instead.

There is still work needed for the VF=1 case because the vplan cost
model is not printed out in this case. I've not attempted to fix those
in this patch.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/no_vector_instructions.ll:
```diff
@@ -6,7 +6,6 @@ target triple = "aarch64--linux-gnu"
 
 ; CHECK-LABEL: all_scalar
 ; CHECK:       LV: Found scalar instruction: %i.next = add nuw nsw i64 %i, 2
-; CHECK:       LV: Found an estimated cost of 1 for VF 2 For instruction: %i.next = add nuw nsw i64 %i, 2
```
Yeah possibly - good suggestion. I'll double check to see if we generate output when no vector instructions are generated.

david-arm replied:
It looks like with the new cost model when we know we won't generate any vector instructions we don't even attempt to calculate a cost.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/scalable-vectorization-cost-tuning.ll:
```diff
@@ -1,54 +1,59 @@
-; VF-4: <4 x i32>
-; VF-VSCALE4: <16 x i32>
+; VF-16: <16 x i8>
+; VF-VSCALE16: <vscale x 16 x i8>
```
I think this test was falling foul of the difference in max vector bandwidth behaviour between NEON and SVE, but perhaps the behaviour is now the same on the latest version of LLVM?

The problem I had was for NEON we'd choose a maximum VF of 16 on neoverse-v1 vs a max of VF=vscale x 4, due to the zext from i8 to i32. This didn't matter for the legacy cost model because the output was divided by the VF, i.e.

```
NEOVERSE-V1: LV: Vector loop of width vscale x 4 costs: 1 (assuming a minimum vscale of 2).
```

That meant we could effectively test for the vscale tuning value, since the final cost = absolute cost / estimated VF. With the new cost model it only prints out the absolute cost per VF so the only way to detect that vscale is having an effect is to check for the chosen VF. But with the old version of the test when NEON is maximising the bandwidth and SVE isn't we end up choosing VF=16.

I wasn't sure if it was worth modifying the new cost model output to show the per-lane cost or not. @fhahn any thoughts?

Another option is to actually get the vectoriser to print out what it estimates the runtime VF to be?

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/type-shrinkage-zext-costs.ll:
```diff
@@ -8,15 +8,14 @@ target triple = "aarch64-unknown-linux-gnu"
-; CHECK-COST: LV: Found an estimated cost of 1 for VF vscale x 2 For instruction:   %conv = zext i8 %0 to i32
-; CHECK-COST: LV: Found an estimated cost of 1 for VF vscale x 4 For instruction:   %conv = zext i8 %0 to i32
-; CHECK-COST: LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %conv = zext i8 %0 to i32
+; CHECK-COST: Cost of 1 for VF 2: WIDEN-CAST ir<%conv> = zext  ir<%0> to i16
```
Ah good point. Looks like I was over eager in deleting the VF=1 case. Yes, at the moment we still use the legacy cost model to calculate the scalar case. I imagine this will get simplified at some point?

david-arm commented on llvm/test/Transforms/LoopVectorize/X86/redundant-vf2-cost.ll:
```diff
@@ -4,11 +4,11 @@
+; CHECK: Cost of 1 for VF 2: WIDEN store vp<%5>, ir<%val>
+; CHECK-NOT: Cost of 1 for VF 2: WIDEN ir<%0> = load vp<%4>
+; CHECK-NOT: Cost of 1 for VF 2: WIDEN store vp<%5>, ir<%val>
+; CHECK: Cost for VF 2: 5
```
Yeah you're right. It's probably better to make it less specific given the CHECK lines have to be hand-written.

david-arm reviewed (commented):
Thanks for the comments @davemgreen!

fhahn commented:
Thank you for adjusting the tests that check the cost model! 

There still a few cases we fall back to legacy costs in the VPlan cost model, which would be good to be protected by the assertion until those are completed. I think the biggest one is finishing computing costs for some reductions  in VPlan directly, which @ElvisWang123 is working on in https://github.com/llvm/llvm-project/pull/113903

davemgreen commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7382,7 +7382,15 @@ InstructionCost LoopVectorizationPlanner::cost(VPlan &Plan,
+    if (std::optional<unsigned> VScale = getVScaleForTuning(OrigLoop, TTI))
+      EstimatedWidth *= *VScale;
+  LLVM_DEBUG(dbgs() << "Cost for VF " << VF << ": " << Cost
+                    << " (Estimated cost per lane: " << (Cost / EstimatedWidth)
```
Could we make this a float? The cost calc will consider fractions I believe (although it will do it the other way around).

david-arm replied:
I've tried to do this, although it's not print with LLVM's ostreams. :)

david-arm committed:
[LoopVectorize][NFC] Rewrite tests to check output of vplan cost model

Currently it's very difficult to improve the cost model for tail-folded
loops because as soon as you add a VPInstruction::computeCost function
that adds the costs of instructions such as VPInstruction::ActiveLaneMask
and VPInstruction::ExplicitVectorLength the assert in
LoopVectorizationPlanner::computeBestVF fails for some tests. This is
because the VF chosen by the legacy cost model doesn't match the vplan
cost model. See PR #90191. This assert is currently making it difficult
to improve the cost model.

Hopefully we will be in a position to remove the assert soon, however
in order to do that we have to fix up a whole bunch of tests that rely
upon the legacy cost model output. I've tried my best to update
these tests to use vplan output instead.

There is still work needed for the VF=1 case because the vplan cost
model is not printed out in this case. I've not attempted to fix those
in this patch.

davemgreen commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7403,7 +7404,23 @@ InstructionCost LoopVectorizationPlanner::cost(VPlan &Plan,
+    std::string CostString;
+    raw_string_ostream CostStream(CostString);
+    double CostPerLane = double(*Cost.getValue()) / EstimatedWidth;
+    llvm::write_double(CostStream, CostPerLane, FloatStyle::Fixed, 1);
```
Could this make use of `dbgs() << format("%1.f", CostPerLane)`?

david-arm replied:
Hmm, I was looking for something like that but couldn't find it. Thanks - I'll give it a try!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7403,7 +7404,20 @@ InstructionCost LoopVectorizationPlanner::cost(VPlan &Plan,
-  LLVM_DEBUG(dbgs() << "Cost for VF " << VF << ": " << Cost << "\n");
+#ifndef NDEBUG
+  unsigned EstimatedWidth = VF.getKnownMinValue();
+  if (VF.isScalable())
+    if (std::optional<unsigned> VScale = getVScaleForTuning(OrigLoop, TTI))
+      EstimatedWidth *= *VScale;
```
It would probably be convenient to have a simple helper for that as we do this in more and more places now?

davemgreen reviewed (approved):
Thanks for the updates - the Arm/AArch64 tests LGTM.

RKSimon reviewed (approved):
X86 LGTM

david-arm commented:
Hi @fhahn, I'll merge this patch this morning, then teach the debug code to use the new helper function as part of PR #116247.

fhahn reviewed (approved):
> Hi @fhahn, I'll merge this patch this morning, then teach the debug code to use the new helper function as part of PR #116247.

Sounds good, thought originally you wanted to merge #116247 first, but either order is fine

david-arm commented:
> > Hi @fhahn, I'll merge this patch this morning, then teach the debug code to use the new helper function as part of PR #116247.
> 
> Sounds good, thought originally you wanted to merge #116247 first, but either order is fine

That was the original plan, but then I had trouble rebasing #116247 and that slowed things down. Plus I'm a bit nervous about delaying this patch for much longer otherwise I fear I'll forever be fixing up conflicts or failing tests. :)

fhahn commented:
@david-arm is it possible that this patch introduced a failure on the windows bots?

https://buildkite.com/llvm-project/github-pull-requests/builds/121177#01934644-4757-4b17-9233-14fafb1473e7


```
FAIL: LLVM :: Transforms/LoopVectorize/ARM/mve-icmpcost.ll (45628 of 56322)
******************** TEST 'LLVM :: Transforms/LoopVectorize/ARM/mve-icmpcost.ll' FAILED ********************
Exit Code: 1
Command Output (stdout):
--
# RUN: at line 1
c:\ws\src\build\bin\opt.exe -passes=loop-vectorize -debug-only=loop-vectorize -disable-output < C:\ws\src\llvm\test\Transforms\LoopVectorize\ARM\mve-icmpcost.ll 2>&1 | c:\ws\src\build\bin\filecheck.exe C:\ws\src\llvm\test\Transforms\LoopVectorize\ARM\mve-icmpcost.ll
# executed command: 'c:\ws\src\build\bin\opt.exe' -passes=loop-vectorize -debug-only=loop-vectorize -disable-output
# note: command had no output on stdout or stderr
# executed command: 'c:\ws\src\build\bin\filecheck.exe' 'C:\ws\src\llvm\test\Transforms\LoopVectorize\ARM\mve-icmpcost.ll'
# .---command stderr------------
# | C:\ws\src\llvm\test\Transforms\LoopVectorize\ARM\mve-icmpcost.ll:224:10: error: CHECK: expected string not found in input
# | ; CHECK: Cost for VF 8: 26 (Estimated cost per lane: 3.2)
# |          ^
# | <stdin>:826:65: note: scanning from here
# | Cost of 0 for VF 8: EMIT branch-on-count vp<%index.next>, vp<%1>
# |                                                                 ^
# | <stdin>:828:1: note: possible intended match here
# | Cost for VF 8: 26 (Estimated cost per lane: 3.3)
# | ^
# |
# | Input file: <stdin>
# | Check file: C:\ws\src\llvm\test\Transforms\LoopVectorize\ARM\mve-icmpcost.ll
# |
# | -dump-input=help explains the following input dump.
# |
Input was:
# | <<<<<<
# |              .
# |              .
# |              .
# |            821: Cost of 4 for VF 8: WIDEN-SELECT ir<%spec.select.i> = select ir<%2>, ir<%shr>, ir<127>
# |            822: Cost of 2 for VF 8: WIDEN-CAST ir<%conv4> = trunc ir<%spec.select.i> to i8
# |            823: Cost of 0 for VF 8: vp<%8> = vector-pointer vp<%next.gep>.1
# |            824: Cost of 2 for VF 8: WIDEN store vp<%8>, ir<%conv4>
# |            825: Cost of 0 for VF 8: EMIT vp<%index.next> = add nuw vp<%2>, vp<%0>
# |            826: Cost of 0 for VF 8: EMIT branch-on-count vp<%index.next>, vp<%1>
# | check:224'0                                                                     X error: no match found
# |            827: Cost of 0 for VF 8: vector loop backedge
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |            828: Cost for VF 8: 26 (Estimated cost per lane: 3.3)
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | check:224'1     ?                                                 possible intended match
# |            829: Cost of 1 for VF 16: induction instruction %dec = add i32 %blkCnt.012, -1
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |            830: Cost of 0 for VF 16: induction instruction %blkCnt.012 = phi i32 [ %dec, %while.body ], [ %blockSize, %while.body.preheader ]
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |            831: Cost of 0 for VF 16: induction instruction %incdec.ptr = getelementptr inbounds i8, ptr %pSrcA.addr.011, i32 1
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |            832: Cost of 0 for VF 16: induction instruction %pSrcA.addr.011 = phi ptr [ %incdec.ptr, %while.body ], [ %pSrcA, %while.body.preheader ]
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |            833: Cost of 0 for VF 16: induction instruction %incdec.ptr5 = getelementptr inbounds i8, ptr %pDst.addr.010, i32 1
# | check:224'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |              .
# |              .
# |              .
# | >>>>>>
````

Looks like the test expects 3.2 but it was 3.3, possibly due to different rounding?

david-arm commented:
Hi @fhahn, yes that's quite possible. Looks like I'm being punished for trying to produce more accurate reporting. :) I'll put in a fix asap or revert. For now the fix is probably to round to an integer value until I can fix it in a better way.

david-arm commented:
Hmm, it would be a shame to lose the more useful debug output. Perhaps I'll just rewrite the test to only check for the "3." bit.

davemgreen commented:
Sorry for the suggestion, floats are often difficult. I agree it is useful to see the actual value though, not rounded to an integer. Perhaps just make that test less precise if it is really 3.25 so could round either way depending on the environment.

david-arm commented:
Hi @davemgreen, not at all it was a good suggestion and I agree it's worth keeping the more accurate debug output as it's extremely helpful when debugging/understanding vectorisation issues. I'll just fix the tests.
